service/history/api/updateactivityoptions/api.go

```diff
@@ -0,0 +1,266 @@
+	ctx context.Context,
+	request *historyservice.UpdateActivityOptionsRequest,
+	shardContext shard.Context,
+	workflowConsistencyChecker api.WorkflowConsistencyChecker,
+) (resp *historyservice.UpdateActivityOptionsResponse, retError error) {
+	_, err := api.GetActiveNamespace(shardContext, namespace.ID(request.GetNamespaceId()))
+	if err != nil {
+		return nil, err
+	}
+
+	validator := api.NewCommandAttrValidator(
+		shardContext.GetNamespaceRegistry(),
+		shardContext.GetConfig(),
+		nil,
+	)
+
+	response := &historyservice.UpdateActivityOptionsResponse{}
+	err = api.GetAndUpdateWorkflowWithNew(
+		ctx,
+		nil,
+		definition.NewWorkflowKey(
+			request.NamespaceId,
+			request.GetUpdateRequest().WorkflowId,
+			request.GetUpdateRequest().RunId,
+		),
+		func(workflowLease api.WorkflowLease) (*api.UpdateWorkflowAction, error) {
+			mutableState := workflowLease.GetMutableState()
+			return updateActivityOptions(shardContext, validator, mutableState, request, response)
+		},
+		nil,
+		shardContext,
+		workflowConsistencyChecker,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return response, err
+}
+
+func updateActivityOptions(
+	shardContext shard.Context,
+	validator *api.CommandAttrValidator,
+	mutableState workflow.MutableState,
+	request *historyservice.UpdateActivityOptionsRequest,
+	response *historyservice.UpdateActivityOptionsResponse,
+) (*api.UpdateWorkflowAction, error) {
+	if !mutableState.IsWorkflowExecutionRunning() {
+		return nil, consts.ErrWorkflowCompleted
+	}
+	updateRequest := request.GetUpdateRequest()
+	mergeFrom := updateRequest.GetActivityOptions()
+	activityId := updateRequest.GetActivityId()
+
+	ai, activityFound := mutableState.GetActivityByActivityID(activityId)
+
+	if !activityFound {
+		// Looks like ActivityTask already completed as a result of another call.
+		// It is OK to drop the task at this point.
+		return nil, consts.ErrActivityTaskNotFound
+	}
+	mask := updateRequest.GetUpdateMask()
+	if mask == nil {
+		return nil, serviceerror.NewInvalidArgument("UpdateMask is nil")
+	}
+
+	updateFields := util.ParseFieldMask(mask)
+	mergeInto := &activitypb.ActivityOptions{
+		TaskQueue: &taskqueuepb.TaskQueue{
+			Name: ai.TaskQueue,
+		},
+		ScheduleToCloseTimeout: ai.ScheduleToCloseTimeout,
+		ScheduleToStartTimeout: ai.ScheduleToStartTimeout,
+		StartToCloseTimeout:    ai.StartToCloseTimeout,
+		HeartbeatTimeout:       ai.HeartbeatTimeout,
+		RetryPolicy: &commonpb.RetryPolicy{
+			BackoffCoefficient: ai.RetryBackoffCoefficient,
+			InitialInterval:    ai.RetryInitialInterval,
+			MaximumInterval:    ai.RetryMaximumInterval,
+			MaximumAttempts:    ai.RetryMaximumAttempts,
+		},
+	}
+
+	// update activity options
+	err := applyActivityOptions(mergeInto, mergeFrom, updateFields)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate the updated options
+	err = validateActivityOptions(validator, request.NamespaceId, ai, mergeInto)
+	if err != nil {
+		return nil, err
+	}
+
+	// move forward activity version
```

Comment:
Where's the assignment from `mergeInto` to `ai`?

Reply:
ah, in my local changes...
will be in the next rev.